app/src/main/java/ch/epfl/sdp/appart/login/FirebaseLoginService.java

```diff
@@ -130,22 +123,43 @@ private FirebaseUser getCurrentFirebaseUser() {
         return user;
     }
 
-
-    private <T> void addCallbackToTask(Task<T> task, LoginCallback callback) {
-        task.addOnCompleteListener(new LoginOnCompleteListener<>(callback));
+    /**
+     * Converts an AuthResult to a User
+     *
+     * @param result the AuthResult coming from Firebase
+     * @return a user
+     */
+    private User getUserFromAuthResult(AuthResult result) {
+        FirebaseUser user = getCurrentFirebaseUser();
+        return new AppUser(user.getUid(), user.getEmail());
     }
 
-    private static class LoginOnCompleteListener<T> implements OnCompleteListener<T> {
-        private final LoginCallback callback;
-
-        LoginOnCompleteListener(LoginCallback callback) {
-            this.callback = callback;
-        }
-
+    /**
+     * Interface that handles the conversion of a task from one type to another
+     *
+     * @param <FROM> the type we want to convert
+     * @param <TO>   the type we want to get
+     */
+    private interface ConvertTask<FROM, TO> {
+        TO convertTask(FROM arg);
+    }
 
-        @Override
-        public void onComplete(@NonNull Task<T> task) {
-            callback.onRequestCompletion(task.isSuccessful());
-        }
+    /**
+     * Completes the future depending on the task result
+     *
+     * @param task the task whose result will go into the returned future if successful
+     * @param func the function used to convert the task from one type to another
+     * @return a future that contains the task result
+     */
+    private <FROM, TO> CompletableFuture<TO> setUpFuture(Task<FROM> task, ConvertTask<FROM, TO> func) {

```

Comment:
As we talked yesterday, I would use the Naming convention of a uppercase letter for the first letter of a type and the rest as lowercase : From, To. Or even simpler but maybe less informative <F, T>

Reply:
Personally, I like the all caps way and the verbosity but I can change it no problem.